frontend/src/APIClients/BaseGQLClient.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  ApolloClient,
+  ApolloLink,
+  createHttpLink,
+  InMemoryCache,
+  Observable,
+} from "@apollo/client";
+import { onError } from "@apollo/client/link/error";
+import { setContext } from "@apollo/client/link/context";
+import { RetryLink } from "@apollo/client/link/retry";
+import jwt from "jsonwebtoken";
+import AUTHENTICATED_USER_KEY from "../constants/AuthConstants";
+import {
+  setLocalStorageObjProperty,
+  getLocalStorageObjProperty,
+} from "../utils/LocalStorageUtils";
+import { REFRESH } from "./mutations";
+
+const getLocalToken = () =>
+  getLocalStorageObjProperty(AUTHENTICATED_USER_KEY, "accessToken");
+
+const httpLink = createHttpLink({
+  uri: `${process.env.REACT_APP_BACKEND_URL}/graphql`,
+  credentials: "include",
+});
+
+const authFromLocalLink = setContext(async (_, { headers }) => {
+  const accessToken = getLocalToken();
+  return {
+    headers: {
+      ...headers,
+      Authorizaton: accessToken ? `Bearer ${accessToken}` : "",
+    },
+  };
+});
+
+const injectAccessToken = async (operation: any) => {
+  let accessToken = getLocalToken();
+  const decodedToken: any = jwt.decode(accessToken);
+  if (
+    operation.operationName !== "Refresh" &&
+    decodedToken &&
+    decodedToken.exp <= Math.round(new Date().getTime() / 1000)
+  ) {
+    /* eslint-disable  @typescript-eslint/no-use-before-define */
+    const results = await client.mutate({
+      mutation: REFRESH,
+    });
+    /* eslint-disable  @typescript-eslint/no-use-before-define */
+    accessToken = results?.data?.refresh.accessToken;
+    setLocalStorageObjProperty(
+      AUTHENTICATED_USER_KEY,
+      "accessToken",
+      accessToken,
+    );
+  }
+  operation.setContext({
+    headers: { Authorization: `Bearer ${accessToken}` },
+  });
+};
+
+// https://www.apollographql.com/blog/apollo-client/using-apollo-link-to-handle-dependent-queries/
+const accessTokenInjectionLink = new ApolloLink(
+  (operation, forward) =>
+    new Observable((observer) => {
+      let handle: any;
+      Promise.resolve(operation)
+        .then((op) => injectAccessToken(op))
+        .then(() => {
+          handle = forward(operation).subscribe({
+            next: observer.next.bind(observer),
+            error: observer.error.bind(observer),
+            complete: observer.complete.bind(observer),
+          });
+        })
+        .catch(observer.error.bind(observer));
+      return () => {
+        if (handle) handle.unsubscribe();
+      };
+    }),
+);
+
+const refreshDirectionalLink = new RetryLink().split(
+  (operation) => ["Refresh", "Login"].includes(operation.operationName),
+  authFromLocalLink.concat(httpLink),
+  accessTokenInjectionLink.concat(httpLink),
+);
+
+// TODO: Turn off on prod
```

Comment:
How will this be handled in staging environments?

Reply:
Errors will continue being console logged in staging, which is fine since there will be no real planetread data and all logins will be controlled by our team.